vllm/v1/spec_decode/eagle.py

```diff
@@ -143,57 +181,115 @@ def propose(
                 last_hidden_states, hidden_states = ret_hidden_states
         sample_hidden_states = last_hidden_states[last_token_indices]
         logits = self.model.compute_logits(sample_hidden_states, None)
-        draft_token_ids = logits.argmax(dim=-1)
 
         # Early exit if there is only one draft token to be generated.
         if self.num_speculative_tokens == 1:
+            draft_token_ids = logits.argmax(dim=-1)
             # [batch_size, 1]
             return draft_token_ids.view(-1, 1)
 
         # TODO: Currently, MTP module released by deepseek only has
         # one layer. Adapt this code to support multiple layers once
         # there's a multi-layer MTP module.
 
-        # Currently FlashAttention is the only backend that supports
+        # Currently, only FlashAttnetion and TreeAttention support
         # multi-token eagle spec decode. This is because the code below
         # makes assumptions about attn_metadata attributes available.
-        assert isinstance(attn_metadata, FlashAttentionMetadata)
+        assert isinstance(attn_metadata,
+                          (FlashAttentionMetadata, TreeAttentionMetadata))
+
+        # Sample a draft token for each tree child at the root level.
+        draft_token_ids = torch.topk(logits,
+                                     self.child_drafts_per_level[0],
+                                     dim=-1).indices
```

Comment:
Is `torch.topk` with k=1 as fast as `torch.argmax`? Otherwise, can we have a branch doing `argmax` for non-trees and `topk` for trees?